Support plugins being passed as functions from .js

JimPatterson opened the pull request:
This allows for locally defined plugins and for sharing lists of plugins
that are used for linting and for directly passing to markdown-it
without repeating the name of the plugin.

In our environment, we have a couple different sets of plugins for
various uses.  To share the config between rendering and linting,
we moved the plugin list to a list with the need data including the
plugin itself.  This change lets us import that list from both places
and use it.

DavidAnson commented:
Could you please help me understand the scenario better?

Comparing:
- https://github.com/DavidAnson/markdownlint-cli2/pull/6/files#diff-313069ad03456e5a4e9e97203ba26711
- https://github.com/DavidAnson/markdownlint-cli2/blob/main/test/markdownItPlugins/function/.markdownlint-cli2.js

The two are very similar. I understand that you might be able to pull the import out into some common/shared file, but each entry in the `markdownItPlugins` array will need to reference an element from that output, so it seems like someone will need to manually update this options file even after your change?

Maybe a more concrete example from your use case would help me understand the duplication you are avoiding?

Thanks!

JimPatterson commented:
We create a file with all the plugin config that looks something like:

```Javascript

const container = require('markdown-it-container');
const externalLinks = require('markdown-it-external-links');
const tables = require('markdown-it-multimd-table');
// lots more imports for plugins...

// local function for slugifying links
function makeSlugify(prefix = '') {
    return function slugify(slug) {
        // elided for simplicity
    };
}

// local markdown-it plugin to add ids to deflist item to link to glossary
function idsForDeflist(md) {
    const slugify = makeSlugify('markdown-def-');

    function addIds(state) {
        for (const [idx, tok] of state.tokens.entries()) {
            if (tok.type === 'dt_open') {
                const open = tok;
                const body = state.tokens[idx + 1];

                open.attrPush(['id', slugify(body.content)]);
            }
        }
    }

    md.core.ruler.push('add-ids-to-deflist', addIds);
}

const anchorOptions = {
    slugify: makeSlugify('markdown-header-'),
};

const basePlugins = [
    [tables, {
        multiline: true,
        rowspan: true,
        headerless: true,
    }],
    [container, 'info'],
    [container, 'warning'],
    [container, 'success'],
    [externalLinks, {
        externalClassName: null,
        externalRel: 'noopener noreferrer',
        externalTarget: '_blank',
    }],
];

const extendedPlugins = [
    ...basePlugins,
    [footnote],
    [anchor, anchorOptions],
    [deflist],
    [superscript],
    [subscript],
    [idsForDeflist],
];

module.exports = {
    MarkdownIt,
    basePlugins,
    extendedPlugins,
};
```

Then, to render any give markdown, we can import the right list from that file and then do:

```Javascript

for (const plugin of extraPlugins) {
    md.use(...plugin);
}
```

Then, for linting, we'd like to require the right list from the file and do:

```Javascript

module.exports = {
    markdownItPlugins: extendedPlugins,
};
```

The goal is that the rendering and the linting use the same set of plugins configured with the same options.

I guess we could put the name of the plugin instead of the actual required plugin, and then for the render side, require it while building the copy of the list.  Likewise, we could move the local plugins each to their own file so that they could be imported.  That gets a little ugly for us since we use a mono-repo, but still doable.

Thoughts?

DavidAnson commented:
That makes sense and it's pretty elegant. I'd forgotten the format I use is almost directly compatible with what the parser needs. The general concept here is reasonable as well.

I can either get very picky with my PR comments on what you've proposed (fair warning here) or I can build on what you have to make this change myself in maybe a week or so. It's up to you whether you want to deal with me being picky or not. :)

JimPatterson commented:
I am okay with either approach.  I have a thick skin and I am willing to do the changes the way would like them to be done if that is easier for you, or I can wait for you to make the change.

DavidAnson commented:
Cool! I hope to provide the first round of feedback this evening.

DavidAnson commented on markdownlint-cli2.js:
```diff
@@ -41,9 +41,15 @@ const readConfig = (dir, name, otherwise) => {
 
 // Require a module ID with the specified directory in the path
```
Changes here will also affect `customRules` and `outputFormatters`, please add similar test cases for them.

JimPatterson replied:
Since there will be three test cases for the same general concept, maybe we should have a shared term so that the cases are related.  Any thoughts on what that term should be?  For now, I'll stick with the `pre-imported` term I used for the other test case.  It was the best I came up with, but is pretty vague.

DavidAnson replied:
That works for me! Same layout for all is a good suggestion.

JimPatterson replied:
These two test cases sets were structured a little differently than the one for markDownItPulgins.  That on uses an intermediate directory with the scenarios under it.  These two did not use that intermediate directory.  I stayed with the existing pattern present for each scenario.

This did result is duplicating one of the custom rules.  I can move to rules up higher in the tree if you'd rather share the rules among the tests.

DavidAnson commented on markdownlint-cli2.js:
```diff
@@ -41,9 +41,15 @@ const readConfig = (dir, name, otherwise) => {
-  const paths = [ dir, ...require.resolve.paths("") ];
-  const resolved = require.resolve(id, { paths });
-  return require(resolved);
+  // Name of module to import?
```
I know comments are good, but I think this and line 51 are unnecessary.

DavidAnson commented on markdownlint-cli2.js:
```diff
@@ -41,9 +41,15 @@ const readConfig = (dir, name, otherwise) => {
+    const resolved = require.resolve(id, { paths });
+    return require(resolved);
+  }
+
```
If comment is removed, the blank line can go, too.

DavidAnson commented on test/markdownItPlugins/pre-imported/.markdownlint-cli2.js:
```diff
@@ -0,0 +1,18 @@
+
+"use strict";
+
+const inline = require("markdown-it-for-inline");
```
Could you please spell it out like “markdownItForInline”?

DavidAnson commented on test/markdownItPlugins/pre-imported/pre-imported.md:
```diff
@@ -0,0 +1,3 @@
+# Heading
```
This is an identical copy of an existing test file, right?

JimPatterson replied:
Yes, it is.  The structure of the test cases appeared to be a directory with the test case name and the supporting config file and some markdown to run it across.  I took the existing `markdownItPlugins/function` case and duplicated it for this new case.

If you have a better method or would rather the content of this test markdown reflect something else, I'm cool with that.

DavidAnson replied:
What you have is exactly what I want!

DavidAnson commented on README.md:
```diff
@@ -175,6 +175,9 @@ $ markdownlint-cli2 "**/*.md" "#node_modules"
 - The format of this file is a [CommonJS module][commonjs-module] that exports
   the object described above for `.markdownlint-cli2.jsonc`.
 - Other details are the same as for `.markdownlint-cli2.jsonc` described above.
+- But, any place that accepts an takes a `string` with the name of
```
Please move the new line above the previous about “other details”.

I’m thinking something more like:
> Instead of passing a `String` to identify the module name/path to load for `customRules`, `markdownItPlugins`, and `outputFormatters`, the corresponding  `Object` or `Function` can be provided directly.

DavidAnson reviewed (changes requested):
Thanks a lot, I left a few comments on the changes. Please also update the commit message to somethings like “Support passing modules directly in .markdownlint-cli2.js.”.

Appreciate the contribution!

DavidAnson commented on README.md:
```diff
@@ -174,6 +174,9 @@ $ markdownlint-cli2 "**/*.md" "#node_modules"
 - The format of this file is a [CommonJS module][commonjs-module] that exports
   the object described above for `.markdownlint-cli2.jsonc`.
+- Instead of passing a `String` to identify the module name/path to load
+  for `customRules`, `markdownItPlugins`, and `outputFormatters`, the
```
Super nit: Will "for" fit on the previous line? If so, will "corresponding" fit on this line?

JimPatterson replied:
Yep.

DavidAnson commented on test/customRules-pre-imported/.markdownlint-cli2.js:
```diff
@@ -0,0 +1,7 @@
+const anyBlockquote = require("./rules/any-blockquote.js");
```
Please add:

```
// @ts-check

"use strict";
```

DavidAnson commented on test/customRules-pre-imported/rules/any-blockquote.js:
```diff
@@ -0,0 +1,26 @@
+// @ts-check
```
Another exact copy, right?

DavidAnson replied:
Ditto for all the test files.

JimPatterson replied:
For the most part yes, since this is testing the ability to process the object/function directly, I removed all but one rule for the custom rules (copy of `test/customRules/`) and all but one output for the output formatters (copy of `test/outputFormatters-params/`).

DavidAnson commented on test/outputFormatters-pre-imported/.markdownlint-cli2.js:
```diff
@@ -0,0 +1,7 @@
+const formatterJson = require("../../formatter-json");
```
Same comment about ts-check, use strict.

DavidAnson commented on test/outputFormatters-pre-imported.formatter.json:
```diff
@@ -0,0 +1,93 @@
+   "no-trailing-spaces"
+  ],
+  "ruleDescription": "Trailing spaces",
+  "ruleInformation": "https://github.com/DavidAnson/markdownlint/blob/v0.20.4/doc/Rules.md#md009",
```
Please rebase onto `main` branch - "v0.20.4" should be "vX.Y.Z" now - sorry I didn't have it this way originally.

DavidAnson reviewed (changes requested):
This looks great, thanks!! I think all the code is here and looks good. A few more nits this time around, but nothing significant. For the next (final?) update, please squash all commits into one and use the message:
> Support passing modules directly in .markdownlint-cli2.js.

Sorry about needing to rebase - I was deliberate about the changes and don't expect anything to conflict - but if it does, then don't worry about rebasing, just update like it is here - I'll rebase manually when I accept the PR.

JimPatterson commented:
Okay, updated based on your comment, squashed and rebased.

Just a FYI: I noticed that the formatter-json/README.md still has a hard coded `/blob/v.20.4` instead of the `X.Y.Z`.  It is that way in main, so I did not change that.

DavidAnson commented:
That README.md is intentional - it's not used by the tests, so won't break as the version of the library changes.

DavidAnson commented:
Thank you - really appreciate the contribution and attention to detail!
